webapp/tests/Controller/ControllerRolesTest.php

```diff
@@ -0,0 +1,234 @@
+            return [];
+        }
+        $crawler = $this->client->request('GET', $url);
+        $response = $this->client->getResponse();
+        $message = var_export($response, true);
+        $this->assertEquals($statusCode, $response->getStatusCode(), $message);
+        return array_unique($crawler->filter('a')->extract(['href']));
+    }
+
+    /**
+     * Test that having the role(s) gives access to all visible pages.
+     * This test should detect mistakes where a page is disabled when the user has a
+     * certain role instead of allowing when the correct role is there.
+     * @var string[] $combinations
+     * @var string[] $roleURLs
+     */
+    private function verifyAccess(array $combinations, array $roleURLs)
+    {
+        foreach ($combinations as static::$roles) {
+            foreach ($roleURLs as $url) {
+                if (
+                    strpos($url, '/jury/change-contest/') !== false ||
+                    strpos($url, 'activate') !== false ||
+                    strpos($url, 'deactivate') !== false
+                ) {
+                    $this->crawlPage($url, 302);
+                } else {
+                    $this->crawlPage($url, 200);
+                }
+            }
+        }
+    }
+
+    public function getAllPages(array $urlsToCheck)
+    {
+        $done = array();
+        do {
+            $toCheck = array_diff($urlsToCheck,$done);
+            foreach ($toCheck as $url) {
+                if (!$this->urlExcluded($url)) {
+                    $urlsToCheck = array_unique(array_merge($urlsToCheck, $this->crawlPage($url, 200)));
+                }
+                $done[] = $url;
+            }
+        }
+        while (array_diff($done,$urlsToCheck));
+        return $urlsToCheck;
+    }
+
+    /**
+     * Test that having the team role for example is enough to view pages of that role.
+     * This test should detect mistakes where a page is disabled when the user has a
+     * certain role instead of allowing when the correct role is there.
+     * @var string $roleBaseURL The standard endpoint from where the user traverses the website
+     * @var string[] $baseRoles The standard role of the user
+     * @var string[] $optionalRoles The roles which should not restrict the viewable pages
+     * @var boolean $allPages Should all possible pages be visited
+     * @dataProvider provideBasePages
+     */
+    public function testRoleAccess(string $roleBaseURL, array $baseRoles, array $optionalRoles, bool $allPages)
+    {
+        static::$roles = $baseRoles;
+        $this->logIn();
+        $urlsToCheck = $this->crawlPage($roleBaseURL, 200);
+        if ($allPages) {
+            $urlsToCheck = $this->getAllPages($urlsToCheck);
+        }
+        $combinations = $this->roleCombinations($baseRoles, $optionalRoles);
+        $this->verifyAccess($combinations, $urlsToCheck);
+    }
+
+    /** The test needs the data of which role to test with its base endpoint and the other optional roles/
```

Comment:
Can you put this first comment line on the next line, i.e. put `/**` on its own line?

Reply:
Should be fixed

Reply:
https://github.com/DOMjudge/domjudge/pull/867#discussion_r491326657
Should also be fixed, not 100% sure if we do now not check less but for a general method which crawls everything staying safe might be better.